waiter/src/waiter/metrics_sync.clj

```diff
@@ -206,31 +207,46 @@
-                           (fn register-outgoing-request [{:keys [out] :as ws-request}]
-                             (let [ws-request (assoc ws-request :request-id request-id)]
-                               (async/>!! out (encrypt {:dest-router-id router-id
-                                                        :request-id request-id
-                                                        :source-router-id my-router-id
-                                                        :tag :initiate}))
-                               (send router-metrics-agent register-router-ws :router-id->outgoing-ws router-id
-                                     ws-request encrypt router-metrics-agent)))
-                           connect-options)))
-          (-> (preserve-metrics-from-routers router-metrics-state (set/union #{my-router-id} (set (keys router-id->http-endpoint))))
+              new-outgoing-router-ids (->> router-metrics-state
+                                           :router-id->outgoing-ws
+                                           utils/keyset
+                                           (set/difference known-router-ids))]
+          (when (not-empty new-outgoing-router-ids)
+            (cid/cinfo "metrics-router-syncer" "new routers:" new-outgoing-router-ids ", known routers" known-router-ids)
+            (doseq [router-id new-outgoing-router-ids]
+              (let [ws-endpoint (-> (get router-id->http-endpoint router-id)
+                                    (str/replace "http://" "ws://")
+                                    (str "waiter-router-metrics"))
+                    request-id (str "inter-router-metrics-" (utils/unique-identifier))
+                    _ (cid/cinfo request-id "connecting to" router-id "at" ws-endpoint)
+                    initiated-promise (promise)
+                    {:keys [^WebSocket socket]} (ws/connect! websocket-client ws-endpoint
+                                                             (fn register-outgoing-request [{:keys [out]}]
+                                                    (cid/cinfo request-id "successfully connected to" router-id)
+                                                    (async/>!! out (encrypt {:dest-router-id router-id
+                                                                             :request-id request-id
+                                                                             :source-router-id my-router-id
+                                                                             :tag :initiated}))
+                                                    ;; trigger publishing of metrics
+                                                    (deliver initiated-promise :initiated))
+                                                             connect-options)]
+                ;; register outside connect! callback to handle messages on the ctrl channel
+                (let [ws-request {:ctrl (.ctrl socket)
```

Comment:
Can we instead avoid adding a bad request to the agent state?

Reply:
I've changed the code to add a separate listener on the `ctrl` chan. The way the code is we can end up creating two listeners on the `ctrl` chan, but they won't affect correctness as `deregister` is called only when receiving non-nil data on the ctrl chan.